### PR TITLE
modernize lib/rarity-map.js

### PR DESCRIPTION
**Pull request by jimmywarting**
#1443

### CODE REVIEW

File: lib/rarity-map.js [hunk 1]
```diff
@@ -1,121 +1,113 @@
...
+class RarityMap {
+  constructor (torrent) {
+    const self = this
 
-  self._torrent = torrent
-  self._numPieces = torrent.pieces.length
-  self._pieces = []
+    self._torrent = torrent
+    self._numPieces = torrent.pieces.length
+    self._pieces = new Array(self._numPieces).fill(0)
```
Comment (DiegoRBaquero) [inline lib/rarity-map.js]: Any benefits in doing this?

File: lib/rarity-map.js [hunk 2]
```diff
@@ -1,121 +1,113 @@
...
 
 /**
  * Mapping of torrent pieces to their respective availability in the torrent swarm. Used
  * by the torrent manager for implementing the rarest piece first selection strategy.
  */
-function RarityMap (torrent) {
-  var self = this
+class RarityMap {
+  constructor (torrent) {
+    const self = this
```
Comment (DiegoRBaquero) [inline lib/rarity-map.js]: We should stop using self

File: lib/rarity-map.js [hunk 3]
```diff
@@ -1,121 +1,113 @@
...
     }
   }
 
-  if (candidates.length > 0) {
-    // if there are multiple pieces with the same availability, choose one randomly
-    return candidates[Math.random() * candidates.length | 0]
-  } else {
-    return -1
+  destroy () {
+    const self = this
```
Comment (DiegoRBaquero) [inline lib/rarity-map.js]: Same here, no more self

File: lib/rarity-map.js [hunk 4]
```diff
@@ -1,121 +1,113 @@
...
+   *
+   * @param {function} pieceFilterFunc
+   * @return {number} index of rarest piece, or -1
+   */
+  getRarestPiece (pieceFilterFunc) {
+    let candidates = []
+    let min = Infinity
+
+    for (let i = 0; i < this._numPieces; ++i) {
+      if (pieceFilterFunc && !pieceFilterFunc(i)) continue
```
Comment (DiegoRBaquero) [inline lib/rarity-map.js]: Much better removing that trueFn

File: lib/rarity-map.js (hunk 1 above)
Comment (jimmywarting) [reply lib/rarity-map.js]: You probably don't need the initial `.fill(0)` in `new Array(self._numPieces).fill(0)` maybe just `new Array(self._numPieces)` will do

Reason why i did this was b/c `new Array(n)` is faster then writing to a literal array `[]`<br>And using `.fill(0)` is faster then writing to each index

skarmavbild 2018-08-26 kl 22 01 14

also b/c this looks cleaner then
```js
recalculate () {	
  this._pieces.fill(0)
}
```
then
```js
recalculate () {	
  for (let i = 0; i < this._numPieces; ++i) {	
    this._pieces[i] = 0	
  }
}
```

File: lib/rarity-map.js (hunk 4 above)
Comment (jimmywarting) [reply lib/rarity-map.js]: thanks

File: lib/rarity-map.js [hunk 5]
```diff
@@ -1,121 +1,108 @@
...
-  var self = this
-
-  self._torrent = torrent
-  self._numPieces = torrent.pieces.length
-  self._pieces = []
+class RarityMap {
+  constructor (torrent) {
+    this._torrent = torrent
+    this._numPieces = torrent.pieces.length
+    this._pieces = new Array(this._numPieces).fill(0)
```
Comment (DiegoRBaquero) [inline lib/rarity-map.js]: What is the fill here for? If it's not needed I think it will just consume O(n) to fill it and can be removed.

**Pull request merged**